File: leetcode/72. Edit Distance.py
# -*- coding: utf-8 -*-

class Solution(object):

    # 核心代码，传入两个字符串x，y以及六种操作的cost，返回产生最小编辑距离的操作和最小cost
    def get_min_edit_diatance(self, x, y, cost_replace, cost_delete, cost_insert):
        # 一些参数的初始化
        x = list(' ' + x)
        y = list(' ' + y)
        m = len(x) - 1
        n = len(y) - 1
        cost = [[float('inf') for col in range(n+1)] for row in range(m+1)]
        for row in range(1, m+1):
            cost[row][0] = cost_delete * row
        for col in range(1, n+1):
            cost[0][col] = cost_insert * col
        cost[0][0] = 0
        # 开始动态规划
        for i in range(1, m+1):
            for j in range(1, n+1):
                cost[i][j] = cost[i][j - 1] + cost_insert
                if x[i] == y[j] and cost[i][j] > cost[i-1][j-1]:
                    cost[i][j] = cost[i-1][j-1]
                if cost[i][j] > cost[i-1][j-1] + cost_replace:
                    cost[i][j] = cost[i-1][j-1] + cost_replace
                if cost[i][j] > cost[i-1][j] + cost_delete:
                    cost[i][j] = cost[i-1][j] + cost_delete
        return cost[m][n]

    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        cost = self.get_min_edit_diatance(word1, word2, 1, 1, 1)
        return cost

File: leetcode/72. Edit Distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = (int)word1.length();
        int n2 = (int)word2.length();
        vector<vector<int>> dp(n1+1, vector<int>(n2+1, 0));
        for (int i = 1; i <= n1; i ++) {
            dp[i][0] = i;
        }
        for (int i = 1; i <= n2; i ++) {
            dp[0][i] = i;
        }
        for (int i = 1; i <= n1; i ++) {
            for (int j = 1; j <= n2; j ++) {
                int tmp = min(dp[i-1][j] + 1, dp[i][j-1] + 1);
                dp[i][j] = min(tmp, dp[i-1][j-1] + 1);
                // printf("i = %d, j = %d, tmp = %d, dp = %d\n", i, j, tmp, dp[i][j]);
                if (word1[i-1] == word2[j-1]) {
                    dp[i][j] = min(dp[i][j], dp[i-1][j-1]);
                }
            }
        }
        return dp[n1][n2];
    }
};
